src/decay.c

```diff
@@ -175,3 +175,123 @@ decay_maybe_advance_epoch(decay_t *decay, nstime_t *new_time,
 
 	return true;
 }
+
+static inline size_t
+decay_npurge_after_interval(decay_t *decay, size_t interval) {
+	size_t i;
+	uint64_t sum = 0;
+	for (i = 0; i < interval; i++) {
+		sum += decay->backlog[i] * h_steps[i];
+	}
+	for (; i < SMOOTHSTEP_NSTEPS; i++) {
+		sum += decay->backlog[i] *
+		    (h_steps[i] - h_steps[i - interval]);
+	}
+
+	return (size_t)(sum >> SMOOTHSTEP_BFP);
+}
+
+static uint64_t
+arena_decay_compute_purge_interval_impl(tsdn_t *tsdn, decay_t *decay,
+    ecache_t *ecache, const purge_interval_opts_t *opts) {
+	if (malloc_mutex_trylock(tsdn, &decay->mtx)) {
+		/* Use minimal interval if decay is contended. */
+		return opts->min_interval_ns;
+	}
+
+	uint64_t interval;
+	ssize_t decay_time = decay_ms_read(decay);
+	if (decay_time <= 0) {
+		/* Purging is eagerly done or disabled currently. */
+		interval = opts->indefinite_sleep;
+		goto label_done;
+	}
+
+	uint64_t decay_interval_ns = decay_epoch_duration_ns(decay);
+	assert(decay_interval_ns > 0);
+	size_t npages = ecache_npages_get(ecache);
```

Comment:
This belongs in the background thread code; the background thread can get the number of pages in the ecache on its own, and then pass it to the decay function.

With that gone, we don't need the ecache parameter to these functions. And with the locking removed above, we don't need the tsdn_t parameter. So this becomes just a "regular" object that has no knowledge of its existence as part of a larger system; it's a small component that can be tested in isolation.